Add a WSG+iiwa LCM level simulation

sammy-tri opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4583)
<!-- Reviewable:end -->

sammy-tri commented:
+@naveenoid for feature review

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-:-KZYZvzIhVsT5OG380uE:b-psw9ws)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Completed a quick pass.

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/CMakeLists.txt, line 61 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYgZgrJfk8UvOK72F9:-KZYgZgrJfk8UvOK72FA:b-4h0rp3) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/CMakeLists.txt#L61)):*
> ```CMake
>   target_link_libraries(iiwa_wsg_simulation
>     drakeKukaIiwaWorld
>     drakeKukaIiwaArmCommon
> ```

Alpha sort.

---

*[drake/examples/kuka_iiwa_arm/CMakeLists.txt, line 66 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYgbjJLVbkhxNsuX4C:-KZYgbjJLVbkhxNsuX4D:b-4h0rp3) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/CMakeLists.txt#L66)):*
> ```CMake
>     drakeSystemAnalysis
>     drakeSystemControllers
>     drakeLCMSystem2)
> ```

Alpha sort.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 3 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYfm1xUq_xZ10vPY2f:-KZYfm1xUq_xZ10vPY2g:b-9qaynm) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L3)):*
> ```cpp
> /// @file
> ///
> /// Implements a simulation of the KUKA iiwa arm.  Like the driver for the
> ```

BTW, consider adding "with a Schunk WSG 50 gripper attached as an end effector" after "arm"

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 5 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYfwulXvhWRuvdhnl5:-KZYfwulXvhWRuvdhnl6:b-tt7kij) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L5)):*
> ```cpp
> /// Implements a simulation of the KUKA iiwa arm.  Like the driver for the
> /// physical arm, this simulation communicates over LCM using lcmt_iiwa_status
> /// and lcmt_iiwa_command messages. It is intended to be a be a direct
> ```

BTW, are there more messages involved now that there is a gripper involved?

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYfbtPIUfwljSfR482:-KZYfbtPIUfwljSfR483:b-4h0rp3) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L46)):*
> ```cpp
> 
> using schunk_wsg::SchunkWsgTrajectoryGenerator;
> using schunk_wsg::SchunkWsgStatusSender;
> ```

Alpha sort.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 78 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYg6PXc7osJ3p0lsB2:-KZYg6PXc7osJ3p0lsB3:bxtp22t) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L78)):*
> ```cpp
>       "/examples/kuka_iiwa_arm/models/objects/simple_cylinder.urdf");
>   tree_builder->StoreModel(
>       "schunk", "/examples/schunk_wsg/models/schunk_wsg_50.sdf");
> ```

BTW, to be consistent with "iiwa", which is a model name, consider replacing "schunk" with "wsg".

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 93 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYfVqn3NIdcigkQuc3:-KZYfVqn3NIdcigkQuc4:b-1pofmh) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L93)):*
> ```cpp
>   // 0.057m thus giving the surface height (`z`) in world coordinates as
>   // 0.736 + 0.057 / 2.
>   const double kTableTopZInWorld = 0.736 + 0.057 / 2;
> ```

BTW, has the rule-of-three been met here?

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 99 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYfRlj5gzBaVisdep-:-KZYfRlj5gzBaVisdep0:b32vgd) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L99)):*
> ```cpp
>   // are set; the default orientations are used in each case.
>   const Eigen::Vector3d kRobotBase(-0.243716, -0.625087, kTableTopZInWorld);
>   const Eigen::Vector3d kCylinderBase(1 + -0.43, -0.65, kTableTopZInWorld + 0.15);
> ```

80

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 109 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYfNBGXjJBMT81K7KB:-KZYfNBGXjJBMT81K7KC:b-6tgtip) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L109)):*
> ```cpp
> 
>   auto plant = std::make_unique<RigidBodyPlant<T>>(tree_builder->Build());
>   // Contact parameters set arbitrarily.
> ```

BTW, missing "are" before "set".

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 140 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYfDf-K3mtfm0mxIb1:-KZYfDf-K3mtfm0mxIb2:b-7ip87a) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L140)):*
> ```cpp
>     // planner.  It's still not a very good match.  Quantitative
>     // comparisons would require torque control and a more careful
>     // estimation of the model constants such as friction in the
> ```

BTW, I doubt friction is a "constant". That's why it's so hard to model! :cry: 

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 145 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYf7cdS5vvdAl17kFc:-KZYf7cdS5vvdAl17kFd:b228yug) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L145)):*
> ```cpp
>     Eigen::VectorXd iiwa_kd = Eigen::VectorXd::Zero(7);
>     Eigen::VectorXd iiwa_ki = Eigen::VectorXd::Zero(7);
>     // y, p, y, p, y, p, y
> ```

What does this mean?

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 194 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYeonaP7_aO92eQVax:-KZYeonaP7_aO92eQVay:bjcy046) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L194)):*
> ```cpp
>     auto wsg_state_port = wsg_ports.second;
> 
>     // Connect the iiwa's gravity compensation.
> ```

BTW, is this comment correct?

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 284 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYeRGPDEW30ULDuD5g:-KZYeRGPDEW30ULDuD5h:bur99nv) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L284)):*
> ```cpp
>       model->get_plant().get_rigid_body_tree();
> 
>   // Creates and adds LCM publisher for visualization.
> ```

BTW, consider replacing "LCM publisher for visualization." with "a DrakeVisualizer."

---

*[drake/examples/kuka_iiwa_arm/urdf/iiwa14_simplified_collision.urdf, line 259 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYe0b8p6YrrQId22y7:-KZYe0b8p6YrrQId22y8:b-kk3ca5) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/urdf/iiwa14_simplified_collision.urdf#L259)):*
> ```XML
>   <link name="iiwa_link_ee">
>     </link>
>   <frame name="iiwa_frame_ee" link="iiwa_link_ee" xyz="0.09 0 0" rpy="3.14159265359 0 1.57079632679"/>
> ```

BTW, why not spell "end_effector" out?

Also, consider avoiding "Yoda" ordering: "iiwa_end_effector_frame".

Also, consider wrapping this line to fit within 80 characters per line.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-:-KZYgfHcapVLDg3Qn4k_:b-ixyaj3)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/CMakeLists.txt, line 61 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYgZgrJfk8UvOK72F9:-KZbHVOC87pobLBXR5tE:b-a2u8rg) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/CMakeLists.txt#L61)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Alpha sort.
</blockquote></details>

Is this requirement documented somewhere?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4583)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/CMakeLists.txt, line 61 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYgZgrJfk8UvOK72F9:-KZbJjM8t1a0wxwQEgGD:bnxt4fa) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/CMakeLists.txt#L61)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Is this requirement documented somewhere?
</blockquote></details>

I don't think we have a style guide for `CMakeLists.txt`, but to use a C++ analogy:

https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

(Maybe we should create a style guide for these build files.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4583)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: 2 of 5 files reviewed at latest revision, 14 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/CMakeLists.txt, line 61 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYgZgrJfk8UvOK72F9:-KZbZUYwPJg9pQoqz5Di:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/CMakeLists.txt#L61)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I don't think we have a style guide for `CMakeLists.txt`, but to use a C++ analogy:

https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

(Maybe we should create a style guide for these build files.)
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/CMakeLists.txt, line 66 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYgbjJLVbkhxNsuX4C:-KZbZXrvFbAhTS-mlojb:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/CMakeLists.txt#L66)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Alpha sort.
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 3 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYfm1xUq_xZ10vPY2f:-KZbZqOgjyG-TGCkHl2A:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L3)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider adding "with a Schunk WSG 50 gripper attached as an end effector" after "arm"
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 5 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYfwulXvhWRuvdhnl5:-KZbZqmxc-DhZvhzr3wR:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L5)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, are there more messages involved now that there is a gripper involved?
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYfbtPIUfwljSfR482:-KZbZuV3cG3ZdFX_3B9x:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L46)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Alpha sort.
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 78 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYg6PXc7osJ3p0lsB2:-KZb_0G6dQtz2s6XSgYt:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L78)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, to be consistent with "iiwa", which is a model name, consider replacing "schunk" with "wsg".
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 93 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYfVqn3NIdcigkQuc3:-KZb_1gIRKgJ12KRoHsY:b-4bxgvw) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L93)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, has the rule-of-three been met here?
</blockquote></details>

We're at least at two.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 99 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYfRlj5gzBaVisdep-:-KZb_9tTO18-03aobTYB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L99)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

80
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 109 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYfNBGXjJBMT81K7KB:-KZb_HSjKPX5cizvpstw:b9gu2er) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L109)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, missing "are" before "set".
</blockquote></details>

Rephrased

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 140 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYfDf-K3mtfm0mxIb1:-KZb_OilydiwR6UTdVNm:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L140)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, I doubt friction is a "constant". That's why it's so hard to model! :cry: 
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 145 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYf7cdS5vvdAl17kFc:-KZb_TZ5QdCe3-k7HaAU:bz2hyw1) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L145)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

What does this mean?
</blockquote></details>

I thought "yaw" and "pitch".  I didn't actually write it so I removed it instead.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 194 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYeonaP7_aO92eQVax:-KZb_nG3fq5hE5_I1Xw3:b9dfuu3) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L194)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, is this comment correct?
</blockquote></details>

Rephrased.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 284 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYeRGPDEW30ULDuD5g:-KZb_qc0MQzt8UqmeP-j:bss4a2b) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L284)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider replacing "LCM publisher for visualization." with "a DrakeVisualizer."
</blockquote></details>

Removed.

---

*[drake/examples/kuka_iiwa_arm/urdf/iiwa14_simplified_collision.urdf, line 259 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZYe0b8p6YrrQId22y7:-KZba0FThavTAmTEnRyD:b-hklf6w) ([raw file](https://github.com/robotlocomotion/drake/blob/89767d82908c5d20cec58a78d10b17aa04f5b13a/drake/examples/kuka_iiwa_arm/urdf/iiwa14_simplified_collision.urdf#L259)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, why not spell "end_effector" out?

Also, consider avoiding "Yoda" ordering: "iiwa_end_effector_frame".

Also, consider wrapping this line to fit within 80 characters per line.
</blockquote></details>

Fixed line wrapping.  Left the rest as consistent with the rest of the file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4583)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/kuka_iiwa_arm/urdf/iiwa14_simplified_collision.urdf, line 260 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZbenrUUqe6pREA3FxY:-KZbenrVcr90Ew86rddH:bextrws) ([raw file](https://github.com/robotlocomotion/drake/blob/dc135116c450d8d892066e484816cae806bcab66/drake/examples/kuka_iiwa_arm/urdf/iiwa14_simplified_collision.urdf#L260)):*
> ```XML
>   </link>
>   <frame name="iiwa_frame_ee" link="iiwa_link_ee"
> 	 xyz="0.09 0 0" rpy="3.14159265359 0 1.57079632679"/>
> ```

BTW, prefer spaces. 

(I know this is a URDF so our C++ style guide doesn't apply, but by analogy: https://google.github.io/styleguide/cppguide.html#Spaces_vs._Tabs)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4583)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/kuka_iiwa_arm/urdf/iiwa14_simplified_collision.urdf, line 260 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZbenrUUqe6pREA3FxY:-KZbrWAwEcqWzvGqgGot:bbrp084) ([raw file](https://github.com/robotlocomotion/drake/blob/dc135116c450d8d892066e484816cae806bcab66/drake/examples/kuka_iiwa_arm/urdf/iiwa14_simplified_collision.urdf#L260)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, prefer spaces. 

(I know this is a URDF so our C++ style guide doesn't apply, but by analogy: https://google.github.io/styleguide/cppguide.html#Spaces_vs._Tabs)
</blockquote></details>

Done.  (Will figure out how to fix in my editor)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4583)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4583)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 76 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZby6dgZ0ZMyN9C7FnX:-KZby6dh_t5cS6ZJfzRP:b-ji1tzf) ([raw file](https://github.com/robotlocomotion/drake/blob/00daac8835f0949d5cbb3bc175229c5f561280fa/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L76)):*
> ```cpp
>       "table",
>       // NOLINTNEXTLINE(whitespace/line_length)
>       "/examples/kuka_iiwa_arm/models/table/extra_heavy_duty_table_surface_only_collision.sdf");
> ```

BTW, consider removing the NOLINT and splitting the above line into two:

```
      "/examples/kuka_iiwa_arm/models/table/"
      "extra_heavy_duty_table_surface_only_collision.sdf");
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4583)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
Reviewed 2 of 5 files at r1, 2 of 3 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 82 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZgi76UJFzyu7XugmGh:-KZgi76VLEsLpa1JQWCT:b-igy9le) ([raw file](https://github.com/robotlocomotion/drake/blob/00daac8835f0949d5cbb3bc175229c5f561280fa/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L82)):*
> ```cpp
>   tree_builder->StoreModel(
>       "wsg", "/examples/schunk_wsg/models/schunk_wsg_50.sdf");
> ```

Maybe include a brief comment here on the world to be constructed..something like, "2 fixed tables are added to the world and a cylinder is placed on one of them while a Kuka IIWA arm is fixed on the other".

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 98 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZghwURdrNpHr7ZkmH5:-KZghwURdrNpHr7ZkmH6:b-sfdqwn) ([raw file](https://github.com/robotlocomotion/drake/blob/00daac8835f0949d5cbb3bc175229c5f561280fa/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L98)):*
> ```cpp
>   const double kTableTopZInWorld = 0.736 + 0.057 / 2;
> 
>   // The positions of the iiwa robot, two cylinders and the cuboid are
> ```

I guess this comment needs updating since you only use a single cylinder..in any case a single comment before the table addition would suffice.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 118 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZgjN_GmdVTcAFbqJ_w:-KZgjN_GmdVTcAFbqJ_x:b-7g7gbp) ([raw file](https://github.com/robotlocomotion/drake/blob/00daac8835f0949d5cbb3bc175229c5f561280fa/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L118)):*
> ```cpp
> }
> 
> // TODO(sam.creasey) We should de-duplicate this with kuka_demo.cc.
> ```

PR #4521 will add a diagram header for the current kuka_demo. Perhaps you can modify your TODO to say that this diagram can also be moved to that common diagram factory header?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4583)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
One pass completed. PTAL.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-:-KZgqa1vUKZiGtzIAza_:bk6bcj6)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 76 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZby6dgZ0ZMyN9C7FnX:-K_0-GIZAm8L_zfdasai:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/00daac8835f0949d5cbb3bc175229c5f561280fa/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L76)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider removing the NOLINT and splitting the above line into two:

```
      "/examples/kuka_iiwa_arm/models/table/"
      "extra_heavy_duty_table_surface_only_collision.sdf");
```
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 82 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZgi76UJFzyu7XugmGh:-K_0-V2-G1c8AX4XE7aq:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/00daac8835f0949d5cbb3bc175229c5f561280fa/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L82)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

Maybe include a brief comment here on the world to be constructed..something like, "2 fixed tables are added to the world and a cylinder is placed on one of them while a Kuka IIWA arm is fixed on the other".
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 98 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZghwURdrNpHr7ZkmH5:-K_0-VfWFZFTNMm65nRp:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/00daac8835f0949d5cbb3bc175229c5f561280fa/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L98)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

I guess this comment needs updating since you only use a single cylinder..in any case a single comment before the table addition would suffice.
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 118 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-KZgjN_GmdVTcAFbqJ_w:-K_0090HzIuOpK11rWgb:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/00daac8835f0949d5cbb3bc175229c5f561280fa/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L118)):*
<details><summary><i>Previously, naveenoid (Naveen Kuppuswamy) wrote…</i></summary><blockquote>

PR #4521 will add a diagram header for the current kuka_demo. Perhaps you can modify your TODO to say that this diagram can also be moved to that common diagram factory header?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4583)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
+@jwnimmer-tri for platform review

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-:-K_00cPvfnPtX_YRVN5F:b-ykw85)*
<!-- Sent from Reviewable.io -->

naveenoid commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-:-K_09rX5p_BG55L2SFVI:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
First pass done.

In general, I would like to have test coverage other than "it didn't segfault within 100ms".  Some of this stuff has easily-measurable correctness properties that it would be nice to trust a test to enforce for us, instead of having to launch and interact with a demo to know if we've broken anything.  But maybe that's for another day.

---

Reviewed 2 of 5 files at r1, 1 of 3 files at r2, 1 of 1 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 101 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_0VTTAKT8kR9B9Gddr:-K_0VTTAKT8kR9B9Gdds:bezybo3) ([raw file](https://github.com/robotlocomotion/drake/blob/0a465791aefdcbfcffe0a2630fd9b1615c4dbcd9/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L101)):*
> ```cpp
>   const double kTableTopZInWorld = 0.736 + 0.057 / 2;
> 
>   const Eigen::Vector3d kRobotBase(-0.243716, -0.625087, kTableTopZInWorld);
> ```

Please explain to future maintainers how to understand and/or repair the two numbers here.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 103 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_0VNPuxiVdmmwX8ygo:-K_0VNPuxiVdmmwX8ygp:bx3wgeu) ([raw file](https://github.com/robotlocomotion/drake/blob/0a465791aefdcbfcffe0a2630fd9b1615c4dbcd9/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L103)):*
> ```cpp
>   const Eigen::Vector3d kRobotBase(-0.243716, -0.625087, kTableTopZInWorld);
>   const Eigen::Vector3d kCylinderBase(
>       1 + -0.43, -0.65, kTableTopZInWorld + 0.15);
> ```

BTW Explain why the cylinder is 15cm above the tabletop?  Does it start the simulation by falling a little?

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 161 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_0WTo0dntXvTvB_h69:-K_0WTo0dntXvTvB_h6A:b-677tb8) ([raw file](https://github.com/robotlocomotion/drake/blob/0a465791aefdcbfcffe0a2630fd9b1615c4dbcd9/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L161)):*
> ```cpp
>         iiwa_kp, iiwa_ki, iiwa_kd, &builder);
>     auto iiwa_control_port = iiwa_ports.first;
>     auto iiwa_state_port = iiwa_ports.second;
> ```

The prior three auto's are overly obscuring of what's going on under the hood.  The types should be more manifest -- probably written out entirely, though perhaps the first could stay auto and the second two written our, or using std::tie, or something.

Furthermore, though, I don't think FooPortDescriptors should be passed around by-value; they are virtual and thus prone to slicing.

The same comment applies to the similar wsg code later on in this method.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 169 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_0XWdvX3xglr1F95iP:-K_0XWdwsFXcwDwiFtsU:bvk99ln) ([raw file](https://github.com/robotlocomotion/drake/blob/0a465791aefdcbfcffe0a2630fd9b1615c4dbcd9/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L169)):*
> ```cpp
>         index_map.at("left_finger_sliding_joint");
>     const int position_index = plant_->FindInstancePositionIndexFromWorldIndex(
>         wsg_instance_id, left_finger_position_index);
> ```

BTW This seems like an overly-roundabount way to find an instance offset index -- going through to the word frame and then backing out the instance base.  It seems like the multibody plant (or tree) should be able to answer this question directly.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 203 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_0YCdqKd49V7KWvJf1:-K_0YCdrWjMFsy3dw59x:bm59mfa) ([raw file](https://github.com/robotlocomotion/drake/blob/0a465791aefdcbfcffe0a2630fd9b1615c4dbcd9/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L203)):*
> ```cpp
>     drake::parsers::urdf::AddModelInstanceFromUrdfFile(
>         drake::GetDrakePath() +
>         "/examples/kuka_iiwa_arm/urdf/iiwa14_simplified_collision.urdf",
> ```

Brittle duplication of string constant higher up in this file.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 220 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_0YrICoeBz-Y5UGhy5:-K_0YrICoeBz-Y5UGhy6:borpokc) ([raw file](https://github.com/robotlocomotion/drake/blob/0a465791aefdcbfcffe0a2630fd9b1615c4dbcd9/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L220)):*
> ```cpp
>               iiwa_tree_.get_num_velocities()});
>     builder.Connect(zero_source->get_output_port(),
>                     input_mux->get_input_port(1));
> ```

BTW On first read, I didn't understand the `1` here at all.  Then I saw that the vector constructor was `{}` and not `()`.  That is very subtle!

For one idea, consider commenting that input_mux.get_input_port(0) is going to be exported, since that line is way lower in this method.

Or maybe declare the mux first, connect its output, and _then_ zero its velocity input?

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 227 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_0_hvcKFgo4ubBBWhE:-K_0_hvd3BrrVPGORp3r:blns41f) ([raw file](https://github.com/robotlocomotion/drake/blob/0a465791aefdcbfcffe0a2630fd9b1615c4dbcd9/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L227)):*
> ```cpp
> 
>     // Split the input state into two signals one with the positions and one
>     // with the velocities.
> ```

BTW By this point in the read-through, I'm wondering why positions and velocities are ever part of the same vector (port)?  It seems like most use cases want to treat them separately, or at least wouldn't be much worse by twaining them.

At a _minimum_, it seems like there's at least missing multibody-specific "qv -> q,v' and "q,v -> qv" Systems for mux and demuxing, that know about the RBT and/or RBP indexing, and encapsulate this splitting/merging.

---

*[drake/examples/kuka_iiwa_arm/iiwa_world/world_sim_tree_builder.h, line 94 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_0SP_xQQ0rjPKjKN9k:-K_0SP_xQQ0rjPKjKN9l:b7k5z3z) ([raw file](https://github.com/robotlocomotion/drake/blob/0a465791aefdcbfcffe0a2630fd9b1615c4dbcd9/drake/examples/kuka_iiwa_arm/iiwa_world/world_sim_tree_builder.h#L94)):*
> ```objc
>   }
> 
>   const RigidBodyTree<T>& tree() const { return *rigid_body_tree_; }
> ```

Missing API doc of `@pre Build() has not been called yet`.

Consider adding DRAKE_DEMAND as a sanity check of the @pre.

Also consider moving this method prior to Build, for extra clarity on the same topic.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-:-K_0b5YHLGZXjyCrg2e9:b7vkodw)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
I agree a more detailed test would be nice.  My current plan is that we will (eventually) have tests which create the iiwa+WSG system and perform operations with the combined system and check that an expected result is reached.  At this point, I'm satisfied with this level of coverage to make sure that the test compiles and that no assertions are violated in the construction of the combined system.  I think any breakage which occurs until more complete tests have been developed are the responsibility of the developers working in kuka_iiwa_arm.

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 101 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_0VTTAKT8kR9B9Gddr:-K_13_P9c3Fw1AaCRxjn:bbkb5b0) ([raw file](https://github.com/robotlocomotion/drake/blob/0a465791aefdcbfcffe0a2630fd9b1615c4dbcd9/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L101)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Please explain to future maintainers how to understand and/or repair the two numbers here.
</blockquote></details>

Stolen outright from iiwa_world_demo.cc.  @naveenoid any idea what would be a reasonable explanation here?

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 103 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_0VNPuxiVdmmwX8ygo:-K_14nmv0kwUT4kiMRHy:bw01mur) ([raw file](https://github.com/robotlocomotion/drake/blob/0a465791aefdcbfcffe0a2630fd9b1615c4dbcd9/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L103)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Explain why the cylinder is 15cm above the tabletop?  Does it start the simulation by falling a little?
</blockquote></details>

Reduced to 10cm, which got rid of most of the falling without issues.  Added comment.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 161 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_0WTo0dntXvTvB_h69:-K_16RYplGQqUQxKPXPj:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0a465791aefdcbfcffe0a2630fd9b1615c4dbcd9/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L161)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The prior three auto's are overly obscuring of what's going on under the hood.  The types should be more manifest -- probably written out entirely, though perhaps the first could stay auto and the second two written our, or using std::tie, or something.

Furthermore, though, I don't think FooPortDescriptors should be passed around by-value; they are virtual and thus prone to slicing.

The same comment applies to the similar wsg code later on in this method.
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 203 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_0YCdqKd49V7KWvJf1:-K_174cq4EZXpbkRPHwV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0a465791aefdcbfcffe0a2630fd9b1615c4dbcd9/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L203)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Brittle duplication of string constant higher up in this file.
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 220 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_0YrICoeBz-Y5UGhy5:-K_17XP9onNZpWhHb2_n:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0a465791aefdcbfcffe0a2630fd9b1615c4dbcd9/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L220)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW On first read, I didn't understand the `1` here at all.  Then I saw that the vector constructor was `{}` and not `()`.  That is very subtle!

For one idea, consider commenting that input_mux.get_input_port(0) is going to be exported, since that line is way lower in this method.

Or maybe declare the mux first, connect its output, and _then_ zero its velocity input?
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 227 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_0_hvcKFgo4ubBBWhE:-K_11jk4NgbqaK8c6nhm:b-m9o7ql) ([raw file](https://github.com/robotlocomotion/drake/blob/0a465791aefdcbfcffe0a2630fd9b1615c4dbcd9/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L227)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW By this point in the read-through, I'm wondering why positions and velocities are ever part of the same vector (port)?  It seems like most use cases want to treat them separately, or at least wouldn't be much worse by twaining them.

At a _minimum_, it seems like there's at least missing multibody-specific "qv -> q,v' and "q,v -> qv" Systems for mux and demuxing, that know about the RBT and/or RBP indexing, and encapsulate this splitting/merging.
</blockquote></details>

Historical convention?  When I added per-instance input and output ports, I was going for an implementation which could be used with a minimum of modification to code which was written to use the traditional "all state" output port.  If we want to make further changes, I think opening an issue would be more appropriate than doing so in this PR.

---

*[drake/examples/kuka_iiwa_arm/iiwa_world/world_sim_tree_builder.h, line 94 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_0SP_xQQ0rjPKjKN9k:-K_12kVo0oQTMV1pRyYD:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0a465791aefdcbfcffe0a2630fd9b1615c4dbcd9/drake/examples/kuka_iiwa_arm/iiwa_world/world_sim_tree_builder.h#L94)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Missing API doc of `@pre Build() has not been called yet`.

Consider adding DRAKE_DEMAND as a sanity check of the @pre.

Also consider moving this method prior to Build, for extra clarity on the same topic.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-:-K_12rDcCaeeYjU5-_Yb:bf0bnbd)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending open discussions, CI, etc.

---

Reviewed 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 227 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_0_hvcKFgo4ubBBWhE:-K_19LMjk4d5gYOe3bnu:b3arnxy) ([raw file](https://github.com/robotlocomotion/drake/blob/0a465791aefdcbfcffe0a2630fd9b1615c4dbcd9/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L227)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Historical convention?  When I added per-instance input and output ports, I was going for an implementation which could be used with a minimum of modification to code which was written to use the traditional "all state" output port.  If we want to make further changes, I think opening an issue would be more appropriate than doing so in this PR.
</blockquote></details>

Yeah.  I am happy with a TODO / issue / etc.  ISTR that Automotive had (has) the same pain, so a common solution would be great.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 65 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_18x-HE-bnueVlugUJ:-K_18x-HE-bnueVlugUK:b-203c40) ([raw file](https://github.com/robotlocomotion/drake/blob/836d3b374af290ee7ed46a8ec7362f4255f8d1a7/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L65)):*
> ```cpp
> using systems::Simulator;
> 
> const char* kIiwaUrdf =
> ```

BTW Consider `const char* const`.  I can never remember if that's a rule or not, but _shrug).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-:-K_19_mmjyUR7WnlxSsS:b-i6mtcb)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 101 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_0VTTAKT8kR9B9Gddr:-K_19iZUgLFJ0j4Wv1CW:b3axl0r) ([raw file](https://github.com/robotlocomotion/drake/blob/0a465791aefdcbfcffe0a2630fd9b1615c4dbcd9/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L101)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

Stolen outright from iiwa_world_demo.cc.  @naveenoid any idea what would be a reasonable explanation here?
</blockquote></details>

FYI for this PR, I'd be fine with a comment saying you stole them and citing the other copied code as the source of truth, with a second PR follow-up (possibly during de-duplication) fixing the where-the-numbers-came-from documentation.  (I had thought this PR was the origin of them.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4583)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 101 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_0VTTAKT8kR9B9Gddr:-K_1BA5PhH7yQORzr9ZK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0a465791aefdcbfcffe0a2630fd9b1615c4dbcd9/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L101)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI for this PR, I'd be fine with a comment saying you stole them and citing the other copied code as the source of truth, with a second PR follow-up (possibly during de-duplication) fixing the where-the-numbers-came-from documentation.  (I had thought this PR was the origin of them.)
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 227 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_0_hvcKFgo4ubBBWhE:-K_1BWx6zaxiLyBPHHGa:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0a465791aefdcbfcffe0a2630fd9b1615c4dbcd9/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L227)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Yeah.  I am happy with a TODO / issue / etc.  ISTR that Automotive had (has) the same pain, so a common solution would be great.
</blockquote></details>

Done.

---

*[drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc, line 65 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4583#-K_18x-HE-bnueVlugUJ:-K_1Ay_Ftftdv7hCbZzQ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/836d3b374af290ee7ed46a8ec7362f4255f8d1a7/drake/examples/kuka_iiwa_arm/iiwa_wsg_simulation.cc#L65)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Consider `const char* const`.  I can never remember if that's a rule or not, but _shrug).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4583)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4583)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@drake-jenkins-bot linux-gcc-bazel-experimental please
@drake-jenkins-bot linux-xenial-gcc-bazel-experimental please
